Add font utilities to derive list of available fonts from font dirs

adam-james-v opened the pull request:
Epic: #22582


This PR adds two defsettings to facilitate Enterprise users' ability to set the font for their instance. The
'available-fonts' can be grabbed by the frontend to populate the dropdown menu. The 'application-font' setting can be
written by the frontend to save the chosen font.

The 'contract' between the frontend and backend is as follows:

`GET /api/settings/available-fonts` -> `("Lato", "Merriweather", "Montserrat", "Noto Sans JP", ...)`
`PUT /api/settings/application-font` -> expects a String
`PUT /api/settings/application-font` -> returns currently set Font's display name eg. `Noto Sans JP`

adam-james-v committed:
Add font utilities to derive list of available fonts from font dirs

This PR adds two defsettings to facilitate Enterprise users' ability to set the font for their instance. The
'available-fonts' can be grabbed by the frontend to populate the dropdown menu. The 'application-font' setting can be
written by the frontend to save the chosen font.

npfitz committed:
Adding some fonts and font selector

adam-james-v committed:
Send  (("dirname", "font name"),) to frontend

Instead of only sending a list of names, send a tuple of `[directory-name, Display Name]`

npfitz committed:
Adding half of the fonts

adam-james-v committed:
Font list derived from font directory/font file names

This is an attempt to derive the 'Display Name' of a font by comparing the directory name to the file names. '_' is
replaced with spaces, but '-' exists in the 'Lato' font directory name, and we may have naming confusion in the
future. Basically, I want to be liberal with what font directory/file names we can receive.

Not certain this is the right approach yet, but it will work and correctly provides names based on the fonts available.

npfitz committed:
Adding remaining fonts

npfitz committed:
Using Default Font family in visual components

npfitz committed:
Removing Noto Sans JP

adam-james-v committed:
Small change to fix some lint warnings

adam-james-v committed:
Small change to fix some linting issues

Merge branch 'whitelabelling-users-font-family-font-setting' of https://github.com/metabase/metabase into whitelabelling-users-font-family-font-setting

adam-james-v committed:
Move logging

adam-james-v committed:
Alter hardcoded font path

npfitz committed:
Added cypress test

adam-james-v committed:
Change Cypress Test to look for correct text

npfitz committed:
adding visual test

npfitz committed:
fixing cypress test

adam-james-v committed:
Added unit tests for fonts.clj, and simplified creating display name

The display name approach is simple -> underscores become spaces, and names are split on dashes/first word is taken
from the split.

This might not be robust to future font installations, but will be a good v1

adam-james-v committed:
Merge branch 'whitelabelling-users-font-family-font-setting' of https://github.com/metabase/metabase into whitelabelling-users-font-family-font-setting

adam-james-v committed:
Actually add the unit tests

adam-james-v committed:
Fix lint problems

npfitz commented:
Notes on frontend implemtation:
- I noticed I subbed in the font-family for static-viz components. Perhaps I should revert these changes, as font changes aren't meant to impact here (though, I am curious if this means that they will be rendered with those fonts. Maybe someone knows?)
- All `@font-family` definitions were moved to their own file
- The CSS variable --default-font-family is now set in `App.jsx`
- There are 2 frontend tests written for this. One simply checks that the correct font family is applied on load, and the other is a percy snapshot test

npfitz commented on frontend/src/metabase/admin/settings/components/widgets/SettingSelect.jsx:
```diff
@@ -15,6 +15,9 @@ const SettingSelect = ({
     onChange={e => onChange(e.target.value)}
+    buttonProps={{
+      dataTestId: key,
+    }}
```
This was done to help make testing easier. I did need to adjust 1 cypress test because of this change

dpsutton commented on src/metabase/public_settings.clj:
```diff
@@ -273,6 +274,12 @@
   :type       :json
   :default    {})
 
+(defsetting application-font
```
should this verify the set value against `available-fonts`? See `site-locale`'s setter for prior art.

noahmoss commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+  "Use a font's directory to derive a Display Name.
+
+  Underscores become spaces, and then split on dashes and take the first word.
+  This is done because of the Lato-v16-latin directory name."
```
Curious, is there a reason the directory name couldn't just be changed to `Lato` to simplify the normalization required?

adam-james-v replied:
Hm, good question. Maybe I had a bit of tunnel vision haha. I'll ask around.

npfitz replied:
I think that would be a decent call given that the other fonts aren't that complicated. I'll need to change some URLs on the frontend, but that's easily done

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+            [clojure.tools.logging :as log]
+            [metabase.util.files :as u.files]))
+
+(def font-path
```
Consider making this variable private.

adam-james-v replied:
Good point. I use this in font_test.clj, is there a quick and easy way to use a private def in another namespace?

I know `#'fonts/private-fn` works for functions, but I don't think that'll work for a private def.

noahmoss replied:
You can use `@#'fonts/private-fn`. `#'fonts/private-fn` is the var; deref-ing it with `@` gets its contents

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+(defn- normalize-font-dirname
+  "Use a font's directory to derive a Display Name.
+
+  Underscores become spaces, and then split on dashes and take the first word.
+  This is done because of the Lato-v16-latin directory name."
+  [dirname]
+  (-> dirname
+      (str/split #"-")
+      first
+      (str/replace #"_" " ")))
```
The steps in the description could correspond to the steps in the code.

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+(defn- contains-font-file?
+  [path]
+  ;; todo: expand this to allow other font formats?
+  (some #(str/includes? % ".woff") (u.files/files-seq path)))
```
Shouldn't `str/includes?` really be `str/ends-with?`?

adam-james-v replied:
I did this intentionally since the files are actually `.woff2`, but wanted to also allow `.woff`. 
I could be more strict and do `(str/ends-with? ".woff2")`

metamben replied:
I see. I would match with a regex then.

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+      first
+      (str/replace #"_" " ")))
+
+(defn- contains-font-file?
```
The function returns the font file (if any), so it might be named `contained-font-file` or something similar.

adam-james-v replied:
I'll probably wrap in `(boolean ...)` to give the expected output.

metamben commented on test/metabase/util/fonts_test.clj:
```diff
@@ -0,0 +1,39 @@
+(deftest available-fonts-test
+  (let [fonts (fonts/available-fonts)]
+    (testing "A list of available fonts is returned"
+      (seq fonts))
```
Isn't `is` missing?

adam-james-v committed:
Add setter to application-font to prevent invalid font setting

adam-james-v committed:
Addressing PR review points

- fixed 'normalize-font-dirname' description
- 'contains-font-file?' returns true/false
- 'available-fonts-test' now properly uses 'is'

noahmoss commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,41 @@
+
+(def font-path
+  "font resources path as a String."
+  "./resources/frontend_client/app/fonts/")
```
It would probably be better to avoid hard-coding this as a relative path...usually when we access stuff in `resources/` we use `clojure.java.io/resource` to look up the resource on the classpath.

That returns a `java.net.URL` with the `file:` protocol though so I guess it doesn't play nice with directories...

I guess you could call `resource`, get the URL string with `.toString`, split it on `:`, take the second element and then pass that into `u.files/get-path`. That's still hacky though, maybe there's a better way.

adam-james-v committed:
Make font-path private

adam-james-v committed:
Lint error.

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,45 @@
+       (filter contains-font-file?)
+       (map #(str/replace (str %) font-path ""))
+       (map normalize-font-dirname)
+       (sort-by #(str/lower-case %))))
```
Consider using `(apply sorted-set-by str/lower-case)` here and `(contains? (available-fonts) font)` to implement `available-font?`.

ranquild commented on frontend/src/metabase/app.js:
```diff
@@ -66,6 +66,16 @@ const theme = {
 
+console.log(MetabaseSettings.get("application-font"));
+
+const applicationFontStyles = css`
```
WDYT about moving these styles to somewhere else, e.g. to `styled-components` or even `styled-components/theme` folder?

ranquild commented on frontend/src/metabase/app.js:
```diff
@@ -66,6 +66,16 @@ const theme = {
   space: [4, 8, 16, 32, 64, 128],
 };
 
+console.log(MetabaseSettings.get("application-font"));
```
A leftover

ranquild reviewed (approved):
Small leftovers:
- static-viz shouldn't be changed - it wouldn't work like that
- .Identifier files should most likely be removed
- console.log statements

Apart from this, great work!

npfitz committed:
PR Cleanup

alxnddr reviewed (commented):
The new fonts did not apply to static visualizations:
<img width="634" alt="Screenshot 2022-05-16 at 20 05 01" src="https://user-images.githubusercontent.com/14301985/168635973-af88345b-a582-4ebc-95fb-b94398a8e1a8.png">
